File: solutions/python3/0216.py
class Solution:
    def combinationSum3(self, k: int, n: int) -> List[List[int]]:
        ret = []
        self.dfs(k, n, 1, [], ret)
        return ret

    def dfs(self, k: int, n: int, s: int, curr: List[int], ret: List[List[int]]) -> None:
        if n < 0:
            return
        if n == 0:
            if k == 0:
                ret.append(curr)
                return

        for i in range(s, 10):
            self.dfs(k - 1, n - i, i + 1, curr + [i], ret)

File: solutions/cpp/0216.cpp
class Solution {
public:
    vector<vector<int>> combinationSum3(int k, int n) {
        vector<vector<int>> ans;
        vector<int> curr;
        dfs(k, n, 1, curr, ans);
        return ans;
    }

    void dfs(int k, int n, int s, vector<int>& curr, vector<vector<int>>& ans) {
        if (n < 0) return;
        if (n == 0) {
            if (k == 0) ans.push_back(curr);
            return;
        }

        for (int i = s; i <= 9; i++) {
            curr.push_back(i);
            dfs(k - 1, n - i, i + 1, curr, ans);
            curr.pop_back();
        }
    }
};